General refactoring and automatic config accumulators (#403)
sous-chefs/chef-grafana: DEVELOPING.md

# Developing

> The refactor of the cookbooks in v10 makes some major changes to the underlying method of operation of the cookbook to improve operation and maintainabilty.

Reviewer: ```suggestion
The refactor for version 10 of this cookbook makes some major changes to the underlying method of operation of the cookbook to improve operation and maintainability.
```

  - `load_current_value` has been implemented for all resources
- The `extra_options` hash property has been added to allow the additional of arbitrary configuration options (all config resources)
- Config files are now generated by the `inifile` and `toml-rb` gems so the templates are purely placeholders

## Adding a new configuration item to an existing resource

Adding a new configuration item to an existing resource is as simple as adding the required item as a property to resource representing the section of the configuration file.

## Adding a new configuration section

To add a new configuration section a `config_` resource should be created where the name following the `config_` prefix is the name of the configuration section to represent. Configuration items under the section are then added as properties to this resource.

The configuration resource **must** then inherit from the `_config_base` partial template via the addition of

```ruby
use 'partial/_config_file'
```

At the top of the resource declaration above the first property.

See a pre-existing resource for an example of this pattern.

### Overiding the automatic path

> If the name of the resource cannot match the name of the configuration section or a nested configuration section is in use such as `section.subsection` then the automatic configuration path can be overriden by the addition of the `:resource_config_path_override` method. Due to needed to override the path in both the action class and the outer resource class this method must be added to both sections of the resource declaration.

Reviewer: ```suggestion
If the name of the resource does not match the name of the configuration section or a nested configuration section is in use such as `section.subsection` then the automatic configuration path can be overriden by the use of the `:resource_config_path_override` method. Due to needing to override the path in both the action class and the outer resource class this method must be added to both sections of the resource declaration.
```

> Below the last property add the following with the array contents set to the config Hash path set as you would path to [Hash#dig](https://ruby-doc.org/core-3.0.2/Hash.html#method-i-dig) to create the (nested if required) configuration path.

Reviewer: I'm not sure this sentence makes sense right now. Are you able to reword it please?

```ruby
def resource_config_path_override
  %w(external_image_storage.s3)
end

action_class do
  def resource_config_path_override
    %w(external_image_storage.s3)
  end
end
```

> ### Excluding properties from the configuration

Reviewer: This works for this cookbook, but if we're to extract this into a library this doesn't have great UX.

Author: I presume you mean where you have to declare the method twice? I'm just wondering actually if the check for resource vs action class can be done in the helper and the correct method called off that.

Reviewer: I think what I mean here is having to add the method to the action_class for it to do something isn't super cool. 
I wonder if there's a property we can add. Or maybe I'm just thinking of this from the end user (wrapper cookbook) end which I shouldn't be. 

This property is meant to be for maintainers to add/remove. That said this pattern isn't really in use in any other part of Chef so I think it's caught me off guard slightly!

Author: Yeh it's the latter, you'd only touch this method/hash value when developing the cookbook.
